h/static/scripts/group-forms/hooks/use-group-annotations.tsx

```diff
@@ -10,36 +16,97 @@ export type GroupAnnotationsOptions = {
+
+  /**
+   * A callback to load the next up-to-20-item chunk of annotations, if any.
+   * Useful to progressively load more annotations, as this hook only loads the
+   * first chunk when invoked.
+   */
+  loadNextPage: () => void;
 };
 
 export function useGroupAnnotations({
   filterStatus,
 }: GroupAnnotationsOptions): GroupAnnotationsResult {
   const config = useContext(Config);
-  const [annotations, setAnnotations] = useState<APIAnnotationData[]>([]);
+  const [annotations, setAnnotations] = useState<APIAnnotationData[]>();
   const [loading, setLoading] = useState(false);
   const [error, setError] = useState<string>();
 
-  useEffect(() => {
+  const pageNumberRef = useRef(1);
+  const totalAnnotationsRef = useRef(0);
+
+  /**
+   * Used to cancel currently in-flight request, whether it's the first one or
+   * any subsequent page triggered by calling `loadNextPage`.
+   */
+  const requestController = useRef<AbortController>();
+
+  const loadAnnotationsForCurrentPage = useCallback(() => {
     if (!config?.api.groupAnnotations) {
       throw new Error('groupAnnotations API config missing');
     }
-    const abort = new AbortController();
 
     setLoading(true);
     fetchGroupAnnotations(config.api.groupAnnotations, {
-      signal: abort.signal,
+      signal: requestController.current?.signal,
+      pageNumber: pageNumberRef.current,
       moderationStatus: filterStatus,
     })
-      .then(setAnnotations)
+      .then(({ annotations, total }) => {
+        // Append annotations from the page to current list
+        setAnnotations((prev = []) => [...prev, ...annotations]);
+        totalAnnotationsRef.current = total;
+      })
       .catch((e: any) => setError(e.message))
       .finally(() => setLoading(false));
-
-    return () => abort.abort();
   }, [config?.api.groupAnnotations, filterStatus]);
 
-  return { annotations, loading, error };
+  const loadNextPage = useCallback(() => {
+    if (
+      !loading &&
+      (!annotations || annotations.length < totalAnnotationsRef.current)
+    ) {
+      requestController.current = new AbortController();
+      pageNumberRef.current++;
+      loadAnnotationsForCurrentPage();
+    }
+  }, [annotations, loadAnnotationsForCurrentPage, loading]);
+
+  const prevFilterStatus = useRef(filterStatus);
+  useEffect(() => {
+    // Every time the filter status changes, discard previous list of
+    // annotations and reset the page number
+    if (prevFilterStatus.current !== filterStatus) {
+      setAnnotations(undefined);
+      pageNumberRef.current = 1;
+    }
+
+    prevFilterStatus.current = filterStatus;
+
+    requestController.current = new AbortController();
+    loadAnnotationsForCurrentPage();
+
+    return () => requestController.current?.abort();
+  }, [filterStatus, loadAnnotationsForCurrentPage]);
+
+  return {
+    loading,
+    loadingFirstPage: loading && annotations === undefined,
+    annotations: annotations ?? [],
+    error,
+    loadNextPage,
+  };
```

Comment:
There is quite a lot of logic here that is not really annotation specific. I expect we might need to refactor it into a shared helper in future, similar to how we have utilities in the LMS app for this, or we might be able to factor the LMS and H pieces into a shared library.